Add pathfinding endpoint for train schedule V2

younesschrifi opened the pull request:
close #6740

younesschrifi committed:
editoast: new pathfinding endpoint

Co-authored-by: Florian Amsallem <florian.amsallem@epita.fr>

flomonster commented on editoast/src/redis_utils.rs:
```diff
@@ -46,6 +49,30 @@ impl ConnectionLike for RedisConnection {
+    ) -> Result<Option<T>> {
+        let value: Option<String> = self.get(key).await?;
+        match value {
+            Some(v) => Ok(serde_json::from_str(&v)?),
```
We should transform the `SerdeError` into a `Redis` custom error (we can type it as `Serialize`). This will throw a 500 instead of a 400.

flomonster commented on editoast/src/redis_utils.rs:
```diff
@@ -46,6 +49,30 @@ impl ConnectionLike for RedisConnection {
+        value: &T,
+        seconds: u64,
+    ) -> Result<()> {
+        let str_value = serde_json::to_string(value)?;
```
We should transform the SerdeError into a Redis custom error (we can type it as Serialize). This will throw a 500 instead of a 400.

flomonster commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,366 @@
+        Err(e) => return Ok(e),
+    };
+    // 2) compute path from core
+    let result = extract_location_from_path_items(conn, infra_id, &path_items).await?;
+    let res = match result {
+        Ok(track_offsets) => pathfinding_core(track_offsets),
+        Err(e) => return Ok(e),
+    };
+    let res = res.unwrap();
```
```suggestion
    let pathfinding_result = pathfinding_core(track_offsets)?;
```

flomonster commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,366 @@
+    };
+    let res = res.unwrap();
+    // 3) Put in cache
+    let seconds = chrono::TimeDelta::weeks(1).num_seconds();
```
We can `use chrono::Duration` in headers.

```suggestion
    let seconds = Duration::weeks(1).num_seconds();
```

flomonster commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,366 @@
+    Ok(res)
+}
+
+/// Hash all the path input informations
```
```suggestion
/// Generates a unique hash based on the pathfinding entries.
/// We need to recalculate the path if:
///   - The path entry is different
///   - The infrastructure has been modified
///   - The application has been updated (the algorithm or payloads may have changed)
```

flomonster commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,366 @@
+    format!("{osrd_version}.{infra}.{infra_version}.{hash_path_input}")
+}
+
+struct TrackOffset {
```
```suggestion
#[derive(Debug, Clone, Serialize)]
struct TrackOffset {
```

flomonster commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,366 @@
+    infra_id: i64,
+    path_items: &[PathItemLocation],
+) -> Result<std::result::Result<Vec<Vec<TrackOffset>>, PathfindingResult>> {
+    // Iterate over path_items to collect trigrams, UICs, and operational point IDs
```
```suggestion
```

flomonster reviewed (commented):
Overall it looks good. We should add some documentation on all `pathfinding.rs` functions.

flomonster commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,366 @@
+#[serde(tag = "status", rename_all = "snake_case")]
+pub enum PathfindingResult {
+    OK {
+        blocks: Vec<NonBlankString>,
```
```suggestion
        blocks: Vec<NonBlankString>,
        routes: Vec<NonBlankString>,
```

flomonster commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,367 @@
+/// Path input is described by some rolling stock information
+/// and a list of path waypoints
+#[derive(Deserialize, Clone, Debug, ToSchema, Hash)]
+pub struct PathfindingInput {
```
Missing loading gauge

flomonster commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,367 @@
+
+crate::routes! {
+    "/v2/infra/{infra_id}/pathfinding/blocks" => {post},
+
```
```suggestion
```

flomonster committed:
editoast: fix train schedule schema

- Replace NonBlankString to Identifier
- Fix Identifier implementation of ToSchema

leovalais commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,401 @@
+#[derive(Serialize, Deserialize, Clone, Debug, ToSchema)]
+#[serde(tag = "status", rename_all = "snake_case")]
+pub enum PathfindingResult {
+    OK {
```
```suggestion
    Success {
```

leovalais commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,401 @@
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, ToSchema)]
+pub struct TrackRanges {
```
```suggestion
pub struct TrackRange {
```

leovalais commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,401 @@
+    conn: &mut PgConnection,
+    infra_id: i64,
+    path_items: &[PathItemLocation],
+) -> Result<std::result::Result<Vec<Vec<TrackOffset>>, PathfindingResult>> {
```
Could you make a `type` alias to make that more readable?

leovalais commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,401 @@
+
+    let ops_id = ops_id.iter().map(|obj_id| (infra_id, obj_id.clone()));
+    let (ops_from_ids, _) =
+        OperationalPointModel::retrieve_batch::<_, Vec<_>>(conn, ops_id).await?;
```
Ignoring missing ids seems sketchy, maybe better call retrieve_batch_or_fail and either panic! or return an error.

flomonster replied:
Ideally, we can fail here, but we need to know which waypoint fails (we need the index). The check is performed later when iterating over the waypoints. If the waypoint operation point cannot be found in the hashmap, we return the error.

leovalais replied:
Good point. In this case I'd go for:

```rust
// a check for missing ids is performed later
let ids_to_ops: HashMap<_, _> = OperationalPointModel::retrieve_batch_unchecked::<_, Vec<_>>(conn, ops_id)
         .await?
         .into_iter()
         .map(|op| (op.obj_id.clone(), op))
         .collect();
```

leovalais commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,401 @@
+    Ok(Ok(()))
+}
+
+fn track_offset_from_op(op: &OperationalPointModel) -> Vec<TrackOffset> {
```
Move as `schema::OperationalPoint::track_offset()`?

leovalais commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,401 @@
+        .clone()
+        .into_iter()
+        .map(|el| TrackOffset {
+            track: el.track.0.into(),
```
```suggestion
            track: el.track,
```

leovalais commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,401 @@
+                secondary_code,
+            } => {
+                let mut ops = trigrams_to_ops.get(&trigram.0).cloned();
+                ops = secondary_code_filter(secondary_code, ops);
```
```suggestion
                let ops = secondary_code_filter(secondary_code, ops);
```
shadowing > mut :p

leovalais commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,401 @@
+        let ids = tracks
+            .iter()
+            .map(|track_offset| (infra_id, track_offset.track.0.clone()));
+        let (_, missings): (Vec<_>, _) = TrackSectionModel::retrieve_batch(conn, ids).await?;
```
retrieve_batch_or_fail

younesschrifi replied:
retrieve_batch_or_fail return an `Err` and we want to return an `Ok`

leovalais commented on editoast/src/views/v2/pathfinding.rs:
```diff
@@ -0,0 +1,401 @@
+        .collect()
+}
+
+fn track_offset_from_ops(ops: &[OperationalPointModel]) -> Vec<TrackOffset> {
```
```suggestion
fn track_offsets_from_ops(ops: &[OperationalPointModel]) -> Vec<TrackOffset> {
```

leovalais commented on editoast/openapi.yaml:
```diff
@@ -3848,6 +3923,281 @@ components:
+            type: array
+          status:
+            enum:
+            - o_k
```
🤣

leovalais reviewed (changes requested):
Mostly lgtm, a few changes and cosmetic remarks:

leovalais reviewed (approved):
LGTM
